test/net_http_test.rb

```diff
@@ -0,0 +1,389 @@
+        if !semian_config[sample_env].key?(semian_identifier)
+          semian_config[sample_env]["http_default"]
+        else
+          semian_config[sample_env][semian_identifier]
+        end
+      end
+
+      with_semian_options(semian_options_proc) do
+        Net::HTTP.start("localhost", TOXIC_PORT) do |http|
+          assert_equal semian_config["development"][http.semian_identifier], http.raw_semian_options
+        end
+        Net::HTTP.start("localhost", PORT) do |http|
+          assert_equal semian_config["development"]["http_default"], http.raw_semian_options
+        end
+        assert_equal semian_config["development"]["http_default"],
+                     Semian::NetHTTP.retrieve_semian_options_by_identifier("http_default")
+      end
+    end
+  end
+
+  def test_custom_raw_semian_options_can_disable
+    with_server do # disabled if nil
+      semian_options_proc = proc { nil }
+      with_semian_options(semian_options_proc) do
+        http = Net::HTTP.new("localhost", "#{TOXIC_PORT}")
+        assert_equal false, http.enabled?
+      end
+    end
+
+    with_server do # disabled if key not found
+      semian_config = {}
+      semian_config["development"] = {}
+      semian_config["development"]["http_localhost_#{TOXIC_PORT}"] =
+        {"tickets" => 1,
+         "success_threshold" => 1,
+         "error_threshold" => 3,
+         "error_timeout" => 10}
+      sample_env = "development"
+
+      semian_options_proc = proc do |semian_identifier|
+        semian_config[sample_env][semian_identifier]
+      end
+      with_semian_options(semian_options_proc) do
+        http = Net::HTTP.new("localhost", "#{TOXIC_PORT}")
+        assert_equal true, http.enabled?
+
+        http = Net::HTTP.new("localhost", "#{TOXIC_PORT + 100}")
+        assert_equal false, http.enabled?
+      end
+    end
+  end
+
+  def test_adding_custom_errors_work
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      with_server do
+        with_custom_errors([::OpenSSL::SSL::SSLError]) do
+          http = Net::HTTP.new("localhost", TOXIC_PORT)
+          assert http.resource_exceptions.include?(::OpenSSL::SSL::SSLError)
+        end
+      end
+    end
+  end
+
+  def test_multiple_different_endpoints_and_ports_are_tracked_differently
+    with_semian_options(DEFAULT_SEMIAN_OPTIONS) do
+      ports = [PORT + 100, PORT + 200]
+      ports.each do |port|
+        Semian["http_localhost_#{port}"].reset if Semian["http_localhost_#{port}"]
+        Semian["http_localhost_#{port + 1}"].reset if Semian["http_localhost_#{port + 1}"]
+        Semian.destroy('http_localhost_#{port}')
+        Semian.destroy('http_localhost_#{port + 1}')
+        with_server(port: port, reset_semian_state: false) do
+          with_toxic(upstream_port: port, toxic_port: port + 1) do |name|
+            open_circuit!(toxic_port: port + 1, toxic_name: name)
+            close_circuit!(toxic_port: port + 1)
+          end
+        end
+      end
```

Comment:
This is also a confusing test, I think it could be cleaned up tremendously if we go with the `with_server` model.